### PR TITLE
proxmox_kvm: fix idempotency issue with state=absent

### PR DESCRIPTION
**Pull request by Aversiste**
##### SUMMARY

When the `vmid` parameter is not supplied and the module can only rely on name look-up an early failure can happen if the targeted VM doesn't exist.
In this case a task execution with the parameter `state` set to `absent` will actually fail instead of being considered ok.

This patch introduces a deferred error-checking for non-existent VMs by assigning the value -1 to the `vmid` parameter, allowing the actual verification to be performed in the right code paths.
Is also help to differentiate between a non-existent `vmid` or non-existent VM `name`.

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

proxmox_kvm

##### ADDITIONAL INFORMATION

Previously:

    TASK [ansible-role-proxmox-instance : Remove instance-test]
    changed: [localhost]
    ...
    TASK [ansible-role-proxmox-instance : Remove instance-test]
    fatal: [localhost]: FAILED! => changed=false
      msg: VM instance-test does not exist in cluster.

Now:

    TASK [ansible-role-proxmox-instance : Remove instance-test]
    ok: [localhost]
    ...
    TASK [ansible-role-proxmox-instance : Remove instance-test]
    ok: [localhost]

### CODE REVIEW

**Comment by Aversiste** [comment]
Added missing changelog fragment.

File: changelogs/fragments/811-proxmox-kvm-state-absent.yml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+---
+bugfixes:
+  - proxmox_kvm - Defer error-checking for non-existent VMs in order to fix idempotency of tasks using `state=absent` and properly recognize a success.
```
Comment (felixfontein) [inline changelogs/fragments/811-proxmox-kvm-state-absent.yml:3]:
```suggestion
  - proxmox_kvm - defer error-checking for non-existent VMs in order to fix idempotency of tasks using ``state=absent`` and properly recognize a success (https://github.com/ansible-collections/community.general/pull/811).
```

**Comment by Aversiste** [comment]
The more I think about this patch the more I prefer the idea of recognizing the `force` parameter with `state=absent` in order to shut-down a VM before removing it. I am going to rework this PR.

**Comment by Aversiste** [comment]
Should I add a `minor_changes` in the changelog fragment ?

**Comment by felixfontein** [comment]
Since you are changing the behavior of the `force` option, you probably need to add a `breaking_change` section for it. And it would make sense to do it in a different PR, since breaking changes cannot be backported to stable-1.

In any case, @helldorado @joshainglis @karmab could you please take a look at this?

**Comment by Aversiste** [comment]
I split the PR in two as you suggested.
The `force=yes` portion is now at #849 with a `breaking_changes` key in its changelog fragment.

File: plugins/modules/cloud/misc/proxmox_kvm.py [hunk 2]
```diff
@@ -904,50 +907,53 @@ def main():
...
+    # the cloned vm name or retrieve the next free VM id from ProxmoxAPI.
     if not vmid:
-        if state == 'present' and (not update and not clone) and (not delete and not revert):
+        if state == 'present' and not update and not clone and not delete and not revert:
             try:
                 vmid = get_nextvmid(module, proxmox)
             except Exception as e:
                 module.fail_json(msg="Can't get the next vmid for VM {0} automatically. Ensure your cluster state is good".format(name))
         else:
+            clone_target = name if not clone else clone
```
Comment (felixfontein) [inline plugins/modules/cloud/misc/proxmox_kvm.py:919]: In Python, you can also write:
```suggestion
            clone_target = clone or name
```

File: plugins/modules/cloud/misc/proxmox_kvm.py [hunk 3]
```diff
@@ -1083,10 +1097,10 @@ def main():
         try:
             vm = get_vm(proxmox, vmid)
             if not vm:
-                module.exit_json(changed=False, msg="VM %s does not exist" % vmid)
+                module.exit_json(changed=False)
 
             if getattr(proxmox.nodes(vm[0]['node']), VZ_TYPE)(vmid).status.current.get()['status'] == 'running':
-                module.exit_json(changed=False, msg="VM %s is running. Stop it before deletion." % vmid)
+                module.exit_json(changed=False, msg="VM %s is running. Stop it before deletion or use force=yes." % vmid)
```
Comment (felixfontein) [inline plugins/modules/cloud/misc/proxmox_kvm.py:1103]: This change does not belong in this PR, but in the other.

Comment (Aversiste) [reply plugins/modules/cloud/misc/proxmox_kvm.py:1103]: You are right I forgot to remove this part.

Comment (Aversiste) [reply plugins/modules/cloud/misc/proxmox_kvm.py]: Corrected.

**Comment by felixfontein** [comment]
@helldorado @joshainglis @karmab one last chance to look at this. If you don't reply until tomorrow, I will just merge this.

**Comment by felixfontein** [comment]
Tomorrow passed, so I merged it.

@Aversiste thanks a lot for fixing this!

**Comment by Andersson007** [comment]
@Aversiste thanks for doing this!
BTW we have `team_virt` which is in charge of proxmox_* modules and it's pretty inactive now (ansibullbot pinged the current members by `cc ..` above).
@Aversiste , if you're interested in working with these modules, we could add you in the team.
It means your `shipit` in PRs will count for automerge (usually needs 2 for fixes / minor changes, new modules are merged manually anyway) and you will receive mail notifications related to proxmox_* modules (not many, i think).
If it sounds good to you, let us know.

Thanks!

**Comment by Aversiste** [comment]
@Andersson007 I would be glad to help with these modules.
